### DISCUSSION TITLE
How do I test a handler with a macro?

### DISCUSSION THREAD

**Discussion by ttfkam** [actix/actix-web#1776] [Q&A] [answered]
Given the following:
```rust
#[get("/")]
async fn hello() -> impl Responder {
    HttpResponse::Ok()
      .content_type("text/plain")
      .body("Hello World!")
}
```
Why doesn't this work?
```rust
    #[actix_rt::test]
    async fn test_hello_ok() {
        let resp = hello().await;
        assert_eq!(resp.status(), StatusCode::OK);
    }
```
I get the compiler error
```
error[E0618]: expected function, found `hello`
  --> src/main.rs:64:20
   |
12 | #[get("/")]
   | ----------- `hello` defined here
...
64 |         let resp = hello().await;
   |                    ^^^^^-----
   |                    |
   |                    call expression requires function
```
Obviously the macro is wrapping/mutating, but the docs make no mention of details. How do I unit test these kinds of endpoints?

**Comment by robjtede** [comment, marked as answer]
You can either use manual routing or something like `#[cfg_attr(not(test), get("/"))]` to conditionally wrap.

Edit: alternative is using `test::init_service`.

**Reply by ttfkam** [reply]
That's... not quite terse, is it? Are the HTTP path macros not considered stable or preferred? Feels a bit odd to have the "Request Routing" example on the project homepage and the very first examples in the "Getting Started" section not be testable as presented.

Seems a doc update with a warning may be in order that only manual routing should be used and macro-based routing is only experimental if this is indeed the case.

**Reply by robjtede** [reply]
The alternative is a more involved test set up using test::init_service with a reduced App registering just the handler you want to test (+ any app data and such). This is how most of actix' lib and integration tests are done.
